Typing into Bitbucket comment fields is sometimes dropping characters/sending editing back to the start of the field

rr-codes opened the pull request:
#### 052c7b2a782b865f1922a7c9a691799aa815d306
<pre>
Typing into Bitbucket comment fields is sometimes dropping characters/sending editing back to the start of the field
<a href="https://bugs.webkit.org/show_bug.cgi?id=271842">https://bugs.webkit.org/show_bug.cgi?id=271842</a>
<a href="https://rdar.apple.com/125039472">rdar://125039472</a>

Reviewed by Ryosuke Niwa and Antti Koivisto.

The original implementation of writing suggestions aka inline text predictions relied on using the
same code path as IME marked text has historically used; i.e., inserting the suggestion directly into
the DOM.

While this worked for trivial, simple cases, this proved to be incompatible from a web-compatibility
perspective, since sites were not expecting this type of mutation to ever be happening in the DOM.
Because of the combination of the conflicting selection changes that both the web engine and the site
do, as well as the specific event handlers that a site may have, this resulted in erratic behavior
when inserting and editing text.

To fix, re-implement this feature using a different and significantly safer implementation; instead
of inserting the suggestion into the DON, simply render it as a pseudo-element while it is being suggested,
and only actually insert it into the DOM once the user accepts it. This significantly improves web
compatibility, and also reduces risk in general by not doing unexpected things that the site can detect.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
* Source/WebCore/css/html.css:
(::-internal-writing-suggestions):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::writingSuggestionData const):
(WebCore::Element::setWritingSuggestionData):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::writingSuggestionData const):
(WebCore::ElementRareData::setWritingSuggestionData):
* Source/WebCore/dom/WritingSuggestionData.h: Copied from Tools/WebKitTestRunner/InjectedBundle/Bindings/TextInputController.idl.
(WebCore::WritingSuggestionData::WritingSuggestionData):
(WebCore::WritingSuggestionData::content const):
(WebCore::WritingSuggestionData::offset const):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectedElement):
(WebCore::Editor::removeWritingSuggestionIfNeeded):
(WebCore::Editor::confirmComposition):
(WebCore::Editor::cancelComposition):
(WebCore::Editor::setWritingSuggestion):
(WebCore::Editor::setComposition):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::writingSuggestionsRenderer const):
(WebCore::RenderElement::setWritingSuggestionsRenderer):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCompositionAsync):
(WebKit::WebPageProxy::setWritingSuggestion):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _setMarkedText:underlines:highlights:selectedRange:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setMarkedText):
(WebKit::compositionAnnotations): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetComposition):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Source/WebKit/WebProcess/WebCoreSupport/glib/WebEditorClientGLib.cpp:
(WebKit::WebEditorClient::didDispatchInputMethodKeydown):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setCompositionForTesting):
(WebKit::WebPage::setCompositionAsync):
(WebKit::WebPage::setWritingSuggestion):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame setMarkedText:selectedRange:]):
(-[WebFrame setMarkedText:forCandidates:]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView setMarkedText:selectedRange:]):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TextInputController.idl:
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.cpp:
(WTR::TextInputController::setMarkedText):
(WTR::createCompositionAnnotationData): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.h:

Canonical link: <a href="https://commits.webkit.org/277282@main">https://commits.webkit.org/277282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb00c2471d3e63771bfc5297f1289c7ee74261de

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49868 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43234 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49492 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31696 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23824 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38438 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47766 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23858 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40671 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19750 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41827 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5229 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43569 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42235 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51744 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22212 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18587 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45731 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23487 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44739 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10410 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24274 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5646,4 +5643,17 @@ void Element::addVisibilityAdjustment(OptionSet<VisibilityAdjustment> adjustment
+String Element::writingSuggestion() const
+{
+    if (!hasRareData())
+        return emptyString();
```
This is slightly inconsistent. If we allocate a new rare data, then the writing suggestion will be null string, not empty string. So this should return the null string `{ }` or we should change rare data to initialize `m_writingSuggestion` to the empty string.

rr-codes replied:
indeed, fixed!

pxlcoder commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -260,7 +263,10 @@ class ElementRareData : public NodeRareData {
+    String m_writingSuggestion;
+
     OptionSet<VisibilityAdjustment> m_visibilityAdjustment;
+    bool m_isVisibilityAdjustmentRoot { false };
```
This looks out of place.

rr-codes replied:
fixed

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2105,8 +2105,54 @@ void Editor::selectComposition()
+    RefPtr node = document->selection().selection().end().protectedDeprecatedNode();
+    auto* result = dynamicDowncast<Element>(node.get());
+    if (result)
+        return result;
```
```suggestion
    if (RefPtr element = dynamicDowncast<Element>(node.get()))
        return element;
```

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2216,7 +2267,36 @@ void Editor::setComposition(const String& text, SetCompositionMode mode)
+
+    m_customCompositionAnnotations = {
+        { "NSTextCompletion"_s /* NSTextCompletionAttributeName */, { suggestionRange } },
+    };
```
This seems very wrong to have in the platform-agnostic `Editor`. Also, I thought the AX changes were coming in a separate PR?

whsieh commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -260,7 +263,10 @@ class ElementRareData : public NodeRareData {
+    String m_writingSuggestion;
+
     OptionSet<VisibilityAdjustment> m_visibilityAdjustment;
+    bool m_isVisibilityAdjustmentRoot { false };
```
(I don't think you meant to add this)

rr-codes replied:
fixed

pxlcoder commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -428,20 +428,6 @@ bool Quirks::shouldDisableElementFullscreenQuirk() const
-        || isDomain("godbolt.org"_s)
-        || m_document->url().host().endsWith("officeapps.live.com"_s)
-        || m_document->url().host().endsWith("onedrive.live.com"_s);
-}
```
Let's do this in a separate PR.

rr-codes replied:
fixed

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2105,8 +2105,54 @@ void Editor::selectComposition()
+        return;
+
+    if (CheckedPtr cache = document->existingAXObjectCache())
+        cache->onTextCompositionChange(*selectedElement, CompositionState::Ended, true, text->data(), m_compositionStart, m_isHandlingAcceptedCandidate);
```
This also seems a little strange, but I'm not particularly familiar with AX machinery. I expect it would be preferred to introduce a new, more-specific method to achieve your desired behavior.

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +263,50 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+    }
+
+    if (!renderer.canHaveChildren())
+        return;
```
You can move this up to be the first check.

whsieh commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2216,7 +2253,25 @@ void Editor::setComposition(const String& text, SetCompositionMode mode)
+    Ref document = protectedDocument();
+    document->updateStyleIfNeeded();
+
+    auto inProgress = m_isHandlingAcceptedCandidate;
```
Is `inProgress` supposed to be used here?

whsieh commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2105,8 +2105,40 @@ void Editor::selectComposition()
+{
+    Ref document = protectedDocument();
+
+    RefPtr node = document->selection().selection().end().protectedDeprecatedNode();
```
Is the intent of this code to find the closest ancestor `Element` that contains the current selection? If so, we may want to consider using `containerNode()`, which handles the `BeforeAnchor` and `AfterAnchor` cases correctly.

If this is really about the anchor node, then `protectedAnchorNode()` or `anchorNode()` should suffice.

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2489,4 +2489,14 @@ void RenderElement::layoutIfNeeded()
+void RenderElement::setWritingSuggestionsRenderer(RenderInline& renderer)
+{
+    ensureRareData().writingSuggestionsRenderer = renderer;
+}
```
Does this need to be on RenderElement? (as opposed to a RenderBlock or even RenderBlockFlow?)

rr-codes replied:
I’m not sure; what’s the difference between those three types?

nt1m replied:
Do subclasses of RenderElement that are not RenderBlock / RenderBlockFlow ever get writing suggestions? If not, then you can store the renderers on the subclasses.

rr-codes replied:
I think theoretically any text-like RenderElement can get them

alanbaradlay replied:
what does text-like RenderElement mean?

rr-codes replied:
I guess I meant like RenderBlock, RenderBlockFlow, RenderInline, and RenderText (I think, at least), and not stuff like RenderImage

alanbaradlay replied:
Since this is writing suggestion on text content I think it's safe to have it on RenderBlockFlow and also RenderInline is only valid in the context of RenderBlockFlow. (btw RenderText is not a RenderElement)

rr-codes replied:
ah ok, apologies. will fix!

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +263,50 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+        m_updater.m_builder.attach(*newWritingSuggestionsRendererParent, WTFMove(newWritingSuggestionsRenderer));
+
+        renderer.setWritingSuggestionsRenderer(*newWritingSuggestionsRendererParent.get());
+        m_updater.m_builder.attach(renderer, WTFMove(newWritingSuggestionsRendererParent));
```
do you know if it works inside inline boxes with isolates and/or bidi in general?

rr-codes replied:
I do not know; do you know how I could test?

I know for some cases this does not work (ie, the suggestion doesn’t show up), however this approach is currently “best effort”, since suggestions may not appear anyways if there aren’t any, and this can be improved over time without having to disable the feature.

alanbaradlay replied:
also (though I think I asked about already) suggestion right after initial letter (not even sure if it's a thing) or just with general mutation e.g. programmatic content removal while editing.

rr-codes replied:
> suggestion right after initial letter
I haven’t been able to trigger a suggestion after a single letter 

> programmatic content removal while editing
do you have an example of this?

alanbaradlay replied:
> I do not know; do you know how I could test?

the first step would be just put contenteditable on a dir=rtl block

rr-codes replied:
yup, it works with rtl

anttijk replied:
So the suggestion is always added at the end of the parent. What happens if we are editing in the middle of a text?

rr-codes replied:
Fixed!

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +263,50 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+        writingSuggestionsRenderer->setStyle(WTFMove(newStyle));
+
+        if (auto child = writingSuggestionsRenderer->firstChild())
+            uncheckedDowncast<RenderText>(child)->setText(renderer.element()->writingSuggestion());
```
why is it ok for this downcast to be unchecked?

rr-codes replied:
fixed

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +263,50 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+        auto newWritingSuggestionsRenderer = WebCore::createRenderer<RenderText>(RenderObject::Type::Text, renderer.document(), renderer.element()->writingSuggestion());
+        m_updater.m_builder.attach(*newWritingSuggestionsRendererParent, WTFMove(newWritingSuggestionsRenderer));
+
+        renderer.setWritingSuggestionsRenderer(*newWritingSuggestionsRendererParent.get());
```
There is a confusion here between what is a "writingSuggestionsRenderer" and what is a "writingSuggestionsRendererParent". You should make this consistent, this does not need the extra confusion.

rr-codes replied:
renamed the variables to be more clear

anttijk commented on Source/WebCore/dom/Element.h:
```diff
@@ -783,6 +783,9 @@ class Element : public ContainerNode {
 
+    void setWritingSuggestion(const String&);
+    String writingSuggestion() const;
+
```
Are there going to be multiple elements with writing suggestions at the same time? If not why does this need to be an Element thing rather than a Document level concept (like Editor and FrameSelection)?

rr-codes replied:
how would I access Editor or FrameSelection from within `RenderTreeUpdater::GeneratedContent`?

anttijk replied:
They are both available in Document which can be found via Node::document() for any Node/Element.

rr-codes replied:
I moved the writing suggestion data to Editor

anttijk commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2489,4 +2489,14 @@ void RenderElement::layoutIfNeeded()
+SingleThreadWeakPtr<RenderInline> RenderElement::writingSuggestionsRenderer() const
+{
+    return hasRareData() ? rareData().writingSuggestionsRenderer : nullptr;
+}
```
Again, is this really a per-renderer rather than a one-per-tree thing?

rr-codes replied:
this is a one-per-tree thing; should this RenderInline member also be moved to Editor? or should it stay here?

alanbaradlay replied:
It would be better if we didn't use the renderer to hold on to editing related temporary state(s) if possible.

rr-codes replied:
I need to store the renderer somewhere though so that I can update it later on though, right?

rr-codes replied:
(similar to backdropRenderer)

alanbaradlay replied:
the renderer is part of the tree, and you can surely hold a (weak) reference to somewhere else (in Editor etc).

anttijk reviewed (commented):
The patch is rather big to review, touching many different areas of code. Can this be done more incrementally?

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -442,6 +442,8 @@ std::optional<Inspector::Protocol::CSS::PseudoId> InspectorCSSAgent::protocolVal
     case PseudoId::WebKitScrollbarCorner:
         return Inspector::Protocol::CSS::PseudoId::WebKitScrollbarCorner;
+    case PseudoId::InternalWritingSuggestions:
+        return Inspector::Protocol::CSS::PseudoId::InternalWritingSuggestions;
```
i actually wonder if we want to expose this in Web Inspector 🤔

if it's not something that developers can directly interact with (i.e. creating their own `foo::-internal-writing-suggestion { ... }` CSS declaration) then it may not be useful to show anything about it, but otherwise it kinda seems more like an internal implementation detail that developers might get confused about

if we do really want to expose this then you'd also need to add support for this in `WI.CSSManager.PseudoSelectorNames` and `WI.CSSManager.displayNameForPseudoId`

rr-codes replied:
ya I don’t want to expose this; in this case, should I just do `return { }`?

dcrousso replied:
yeah i think that should work. and also
please remove it from `Source/JavaScriptCore/inspector/protocol/CSS.json` too :)

darinadler commented on Source/WebCore/dom/Element.h:
```diff
@@ -85,6 +85,7 @@ class Text;
 class UniqueElementData;
 class ValidatedFormListedElement;
 class WebAnimation;
+class WritingSuggestionData;
```
Why do we need to add this here? I don’t see any use of this class name in this file. This must be left over from an earlier version of this patch.

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -150,6 +150,7 @@
 #include "WebAnimation.h"
 #include "WebAnimationTypes.h"
 #include "WheelEvent.h"
+#include "WritingSuggestionData.h"
```
Why do we need to add this include? I don’t see any new code in this file. This must be left over from an earlier version of this patch.

darinadler commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -42,6 +42,7 @@
 #include "StylePropertyMap.h"
 #include "StylePropertyMapReadOnly.h"
 #include "VisibilityAdjustment.h"
+#include "WritingSuggestionData.h"
```
Why do we need to add this include? I don’t see any new code in this file. This must be left over from an earlier version of this patch.

darinadler commented on Source/WebCore/dom/WritingSuggestionData.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace WebCore {
+
+class WritingSuggestionData {
```
I would have made this a structure rather than a class since it’s just a data holder.

rr-codes replied:
I originally had this as a struct; however, since I wanted to have r-values for the constructor, I switched it to a class since I usually try to avoid having custom constructors for structs. I will change it back to a structure!

darinadler replied:
You can still use move semantics when initializing the `struct` even without a construdctrir.

darinadler replied:
Don’t mind if you do this as a follow up. I’m worried about delaying this further.

darinadler commented on Source/WebCore/editing/Editor.h:
```diff
@@ -35,6 +35,7 @@
 #include "FrameSelection.h"
 #include "LocalFrame.h"
 #include "PasteboardWriterData.h"
+#include "RenderInline.h"
```
Unfortunate to add this header include.

darinadler commented on Source/WebCore/editing/Editor.h:
```diff
@@ -607,6 +610,16 @@ class Editor : public CanMakeCheckedPtr<Editor> {
+
+    SingleThreadWeakPtr<RenderInline> writingSuggestionRenderer() const { return m_writingSuggestionRenderer; }
+
+    void setWritingSuggestionRenderer(RenderInline& renderer) { m_writingSuggestionRenderer = renderer; }
```
I don’t think we need all these blank lines.

Not sure the return value from `writingSuggestionRenderer()` needs to be a `WeakPtr`. In our coding style I think it can be a raw pointer.

darinadler commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+
+    auto* writingSuggestionData = editor.writingSuggestionData();
+    if (!writingSuggestionData) {
+        destroyWritingSuggestionsIfNeeded();
```
Seems like we should seek a coding style where we don’t need to call this in so many places.

darinadler replied:
Don’t mind if you do this as a follow up. I’m worried about delaying this further.

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+
+        suffixText->setText(suffix);
+    } else {
+        auto newWritingSuggestionsRenderer = WebCore::createRenderer<RenderInline>(RenderObject::Type::Inline, renderer.document(), WTFMove(newStyle));
+        newWritingSuggestionsRenderer->initializeStyle();
+
+        auto writingSuggestionsText = WebCore::createRenderer<RenderText>(RenderObject::Type::Text, renderer.document(), writingSuggestionData->content());
+        m_updater.m_builder.attach(*newWritingSuggestionsRenderer, WTFMove(writingSuggestionsText));
+
+        editor.setWritingSuggestionRenderer(*newWritingSuggestionsRenderer.get());
+        m_updater.m_builder.attach(renderer, WTFMove(newWritingSuggestionsRenderer));
+
+        auto* prefixNode = firstChildText->textNode();
+        if (!prefixNode) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
```
I still have trouble understanding the implications of this. The text in DOM and rendering side will be out of sync so things like text iterators will return surprising results that include the suggestion text. It seems this could still easily break things. I'm going to say r+ to the rendering/style parts to make progress, trusting you will address any issues that will raise. You'll need to find another reviewer for the editing side.

alanbaradlay replied:
I also have trouble understanding why the empty placeholder with shadow DOM approach was never entertained.

alanbaradlay replied:
It'd be great to see some test coverage on this out-of-sync part of the tree getting mutated by JS (i.e. content editing and JS mutating the same time)

rr-codes replied:
I am actively working on write a bunch of test sites, and will create layout tests from them in a follow up PR.

whsieh commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2216,7 +2258,53 @@ void Editor::setComposition(const String& text, SetCompositionMode mode)
+    ASSERT(newText.startsWith(currentText));
+    auto textDelta = newText.substring(currentText.length());
+
+    auto range = document->selection().selection().firstRange().value();
```
What guarantees that the document selection is not none or orphaned here?

rr-codes replied:
fixed

nt1m commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -1287,6 +1287,7 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr<RenderOb
         // From RenderElement
         std::unique_ptr<ReferencedSVGResources> referencedSVGResources;
         SingleThreadWeakPtr<RenderBlockFlow> backdropRenderer;
+        SingleThreadWeakPtr<RenderInline> writingSuggestionsRenderer;
```
This looks unused

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+void RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer(RenderElement& renderer)
+{
+    auto destroyWritingSuggestionsIfNeeded = [&renderer, this]() {
+        if (!renderer.element())
+            return;
+
```
`renderer.element()` must be present at this point, we should ASSERT it, not early-return

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+            m_updater.m_builder.destroy(*writingSuggestionsRenderer);
+    };
+
+    if (!renderer.canHaveChildren())
+        return;
+
+    if (!renderer.element())
+        return;
```
```suggestion
    if (!renderer.canHaveChildren() || !renderer.element())
        return;
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+        return;
+    }
+
+    auto* firstChildText = dynamicDowncast<RenderText>(renderer.firstChild());
+    if (!firstChildText) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
```
How do we end up reaching this? Shouldn't `writingSuggestionData` be cleared in that case? and `destroyWritingSuggestionsIfNeeded` invoked as a result?

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+        writingSuggestionsRenderer->setStyle(WTFMove(newStyle));
+
+        auto* writingSuggestionsText = dynamicDowncast<RenderText>(writingSuggestionsRenderer->firstChild());
+        if (!writingSuggestionsText) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
```
```suggestion
        RELEASE_ASSERT(writingSuggestionsText);
```
I have a slight preference for something like this in terms of sanity.

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+
+        writingSuggestionsText->setText(writingSuggestionData->content());
+
+        auto* suffixText = dynamicDowncast<RenderText>(writingSuggestionsRenderer->nextSibling());
+        if (!suffixText) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
```
```suggestion
        auto* suffixText = dynamicDowncast<RenderText>(writingSuggestionsRenderer->nextSibling());
        RELEASE_ASSERT(suffixText);
```

I have a slight preference for something like this in terms of sanity.

alanbaradlay replied:
I don't think we should crash the process just because we messed up suggestions -instead we should revert the tree to its original state.

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -263,4 +265,97 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
+        editor.setWritingSuggestionRenderer(*newWritingSuggestionsRenderer.get());
+        m_updater.m_builder.attach(renderer, WTFMove(newWritingSuggestionsRenderer));
+
+        auto* prefixNode = firstChildText->textNode();
+        if (!prefixNode) {
+            ASSERT_NOT_REACHED();
+            destroyWritingSuggestionsIfNeeded();
+            return;
+        }
```
Ditto here

```suggestion
        auto* prefixNode = firstChildText->textNode();
        RELEASE_ASSERT(prefixNode);
```

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2105,8 +2107,43 @@ void Editor::selectComposition()
+        return nullptr;
+
+    if (RefPtr element = dynamicDowncast<Element>(node.get()))
+        return element.get();
```
We should make sure this element is editable.

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2105,8 +2107,43 @@ void Editor::selectComposition()
+    if (RefPtr element = dynamicDowncast<Element>(node.get()))
+        return element.get();
+
+    return node->protectedParentElement().get();
```
Ditto.

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2216,7 +2258,53 @@ void Editor::setComposition(const String& text, SetCompositionMode mode)
+
+    if (!currentText.isEmpty()) {
+        SetForScope isInsertingTextForWritingSuggestionScope { m_isInsertingTextForWritingSuggestion, true };
+        insertText(textDelta, nullptr, TextEventInputKeyboard);
```
So this will insert the current writing suggestions? 
Sounds like we want to reflect that in the name of this function.
Maybe setWritingSuggestionAfterAcceptingCurrentWritingSuggestion?

rr-codes replied:
the `insertText` call only inserts the letter that is typed (since the normal way of inserting text doesn’t get called from AppKit in this case; `setMarkedText` does

rr-codes committed:
Typing into Bitbucket comment fields is sometimes dropping characters/sending editing back to the start of the field
https://bugs.webkit.org/show_bug.cgi?id=271842
rdar://125039472

Reviewed by Ryosuke Niwa and Antti Koivisto.

The original implementation of writing suggestions aka inline text predictions relied on using the
same code path as IME marked text has historically used; i.e., inserting the suggestion directly into
the DOM.

While this worked for trivial, simple cases, this proved to be incompatible from a web-compatibility
perspective, since sites were not expecting this type of mutation to ever be happening in the DOM.
Because of the combination of the conflicting selection changes that both the web engine and the site
do, as well as the specific event handlers that a site may have, this resulted in erratic behavior
when inserting and editing text.

To fix, re-implement this feature using a different and significantly safer implementation; instead
of inserting the suggestion into the DON, simply render it as a pseudo-element while it is being suggested,
and only actually insert it into the DOM once the user accepts it. This significantly improves web
compatibility, and also reduces risk in general by not doing unexpected things that the site can detect.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
* Source/WebCore/css/html.css:
(::-internal-writing-suggestions):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::writingSuggestionData const):
(WebCore::Element::setWritingSuggestionData):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::writingSuggestionData const):
(WebCore::ElementRareData::setWritingSuggestionData):
* Source/WebCore/dom/WritingSuggestionData.h: Copied from Tools/WebKitTestRunner/InjectedBundle/Bindings/TextInputController.idl.
(WebCore::WritingSuggestionData::WritingSuggestionData):
(WebCore::WritingSuggestionData::content const):
(WebCore::WritingSuggestionData::offset const):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectedElement):
(WebCore::Editor::removeWritingSuggestionIfNeeded):
(WebCore::Editor::confirmComposition):
(WebCore::Editor::cancelComposition):
(WebCore::Editor::setWritingSuggestion):
(WebCore::Editor::setComposition):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::writingSuggestionsRenderer const):
(WebCore::RenderElement::setWritingSuggestionsRenderer):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCompositionAsync):
(WebKit::WebPageProxy::setWritingSuggestion):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _setMarkedText:underlines:highlights:selectedRange:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setMarkedText):
(WebKit::compositionAnnotations): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetComposition):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Source/WebKit/WebProcess/WebCoreSupport/glib/WebEditorClientGLib.cpp:
(WebKit::WebEditorClient::didDispatchInputMethodKeydown):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setCompositionForTesting):
(WebKit::WebPage::setCompositionAsync):
(WebKit::WebPage::setWritingSuggestion):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame setMarkedText:selectedRange:]):
(-[WebFrame setMarkedText:forCandidates:]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView setMarkedText:selectedRange:]):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TextInputController.idl:
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.cpp:
(WTR::TextInputController::setMarkedText):
(WTR::createCompositionAnnotationData): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.h:

Canonical link: https://commits.webkit.org/277282@main
